### PR TITLE
8298632: [TESTBUG] Add IR checks in jtreg vectorization tests

### PR DESCRIPTION
**Pull request by pfustc**
In JDK-8183390, we introduced a new auto-vectorization testing framework in hotspot jtreg tests. The new framework provides us a much simpler way to add new test cases of auto-vectorizable loops. But the previous patch just added a check of the correctness of the result after vectorization. There is no check about whether the code is vectorized or not.

This patch adds IR checks to verify C2's vectorization ability on test cases inside this framework. With this patch, each test method annotated with `@Test` is verified in two ways. First, it's invoked twice and the return results from the interpreter and C2 compiled code are compared. Second, the count of expected vector IR is checked by the IR framework if the test method has IR rule annotation.

Ideally, we should check IR rules on all platforms. But in practice, the vectorization ability can be quite different on different platforms, or different generations of CPUs of one platform. So in this patch, we only add vectorizable checks for AArch64. Checks for other platforms (such as x86) can still be added later with more CPU feature conditions. We also add some negative rules (or in-vectorizable rules) with `@IR(failOn=...` on cases that should not be vectorized on any platform.

A few more test cases are added within this patch as well.

We tested the new IR rules on below kinds of CPUs.
- AArch64 w/ 512-bit SVE
- AArch64 w/ 128-bit SVE
- AArch64 w/o SVE (NEON only)
- x86

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8298632](https://bugs.openjdk.org/browse/JDK-8298632): [TESTBUG] Add IR checks in jtreg vectorization tests

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [3a87694f](https://git.openjdk.org/jdk/pull/11687/files/3a87694fd14e4acf72f21da3aee0f9b241b706b0)
 * [Jatin Bhateja](https://openjdk.org/census#jbhateja) (@jatin-bhateja - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11687/head:pull/11687` \
`$ git checkout pull/11687`

Update a local copy of the PR: \
`$ git checkout pull/11687` \
`$ git pull https://git.openjdk.org/jdk pull/11687/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11687`

View PR using the GUI difftool: \
`$ git pr show -t 11687`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11687.diff">https://git.openjdk.org/jdk/pull/11687.diff</a>

</details>

### CODE REVIEW

**Review by vnkozlov — CHANGES_REQUESTED** [review]
Good work. But I think you should add x86 IR testing too. It could be done by adding `"sse2", "true"` to `applyIfCPUFeatureOr`. Or high features (avx2, avx512) if required.

You should also consider reducing time to run these tests since you added additional testing. May be lower value of `SIZE`.

**Comment by vnkozlov** [comment]
> Good work. But I think you should add x86 IR testing too. It could be done by adding `"sse2", "true"` to `applyIfCPUFeatureOr`. Or high features (avx2, avx512) if required.

I tested and almost all tests passed if I use "avx2" instead of my original "sse2".

Few tests requires "avx512dq":

ArrayTypeConvertTest.convertDoubleToLong()
ArrayTypeConvertTest.convertFloatToLong()
ArrayTypeConvertTest.convertLongToDouble()
ArrayTypeConvertTest.convertLongToFloat()

BasicLongOpTest.vectorAbs() needs "avx512vl"

**Comment by vnkozlov** [comment]
Don't forget to change `applyIfCPUFeature` to `applyIfCPUFeatureOr`

And to run these tests with different AVX configuration we need to remove `vm.flagless` from `@requires`. `-XX:UseAVX=n` flag is supported by IR testing now.

**Comment by pfustc** [comment]
Hi @vnkozlov ,

Thanks for your review and test. I have made below changes based on your comments.

- Add `avx2` rules for x86 according to your instructions
- Reduce `SIZE` (some test code is updated to align with the reduced `SIZE`)

> And to run these tests with different AVX configuration we need to remove vm.flagless from `@requires`.

Unfortunately, for now we cannot remove `vm.flagless` from `@requires`. As these test methods are also used for correctness check (each test method is invoked twice and the return results from the interpreter and C2 compiled code are compared) and we are using compiler control via WhiteBox API to force these methods running in interpreter and C2. The compiler control won't work if some extra vm option is specified.

**Comment by pfustc** [comment]
BTW, previously I would like to add IR checks for x86 as well. But I don't have various generations of x86 machines to test so I don't have enough confidence to add them. Thanks @vnkozlov for your test effort on avx2. I would appreciate if someone from Intel (maybe @jbhateja) may help verify the x86 rules.

**Comment by danielogh** [comment]
If we want`UseAVX` or other platform-specific flags in annotations, maybe also need [JDK-8297490](https://bugs.openjdk.org/browse/JDK-8297490)

**Comment by vnkozlov** [comment]
> BTW, previously I would like to add IR checks for x86 as well. But I don't have various generations of x86 machines to test so I don't have enough confidence to add them. Thanks @vnkozlov for your test effort on avx2. I would appreciate if someone from Intel (maybe @jbhateja) may help verify the x86 rules.

You can do testing by using `-XX:UseSSE=n -XX:UseAVX=n` flags on modern (avx512) machine.
To bypass `vm.flagless` you can `export TEST_VM_FLAGLESS=true`. That is what I did for testing.

Thank you for explaining  `vm.flagless` issue. In our testing environment we do run these and other Vector API and vectorization tests with different `-XX:UseSSE=n -XX:UseAVX=n` flags settings. That is why I asked about `vm.flagless`.
Anyway, with added `avx` to IR testing filter we will run them on x86 and it is enough for now.

**Review by vnkozlov — APPROVED** [review]
Looks good.

Need second review. Preferable from some one familiar with x86 IR rules.

**Comment by pfustc** [comment]
> If we wantUseAVX or other platform-specific flags in annotations, maybe also need [JDK-8297490](https://bugs.openjdk.org/browse/JDK-8297490)

@danielogh Thanks for letting me know this. As I have discussed with kvn, currently we have to keep `vm.flagless` in these tests. Users cannot select non-default UseSSE/UseAVX values without the `export TEST_VM_FLAGLESS=true` workaround. So, I don't think I need to add platform-specific flags in `@applyIf` for now.

**Comment by pfustc** [comment]
> Looks good.
>
> Need second review. Preferable from some one familiar with x86 IR rules.

Thanks @vnkozlov for the review.

Can I have a second review for x86?

**Comment by pfustc** [comment]
Hi, anyone can help with a second review? I updated copyright years to 2023 in latest commit.

File: test/hotspot/jtreg/compiler/vectorization/runner/ArrayShiftOpTest.java
Comment (jatin-bhateja) [inline test/hotspot/jtreg/compiler/vectorization/runner/ArrayShiftOpTest.java:77]: IR checking is currently being done for STORE node where as the test is written for ROTATE operation.  AVX512F has direct rotate instruction support. Same comment applies to other cases, we should be checking for the exact operation nodes.

File: test/hotspot/jtreg/compiler/vectorization/runner/ArrayCopyTest.java
Comment (jatin-bhateja) [inline test/hotspot/jtreg/compiler/vectorization/runner/ArrayCopyTest.java:78]: For X86 you can replace avx2 with sse for store vector checks.

File: test/hotspot/jtreg/compiler/vectorization/runner/ArrayShiftOpTest.java
Comment (pfustc) [reply test/hotspot/jtreg/compiler/vectorization/runner/ArrayShiftOpTest.java:79]: Thanks for your good suggestion! In my latest commit, IR rules are updated with more exact operations in most of the cases. The ROTATE operation applies for x86 only so I added an additional rule for `avx512f` there.

File: test/hotspot/jtreg/compiler/vectorization/runner/ArrayCopyTest.java
Comment (pfustc) [reply test/hotspot/jtreg/compiler/vectorization/runner/ArrayCopyTest.java]: This is also updated in my last commit. Could you help verify these again?

**Comment by pfustc** [comment]
Hi @jatin-bhateja , x86 IR rules are updated base on your latest guide. Please let me know if we need further updates.

**Comment by jatin-bhateja** [comment]
> Hi @jatin-bhateja , x86 IR rules are updated base on your latest guide. Please let me know if we need further updates.

LGTM

**Review by jatin-bhateja — APPROVED** [review]
Thanks a lot @pfustc!

**Comment by pfustc** [comment]
Thanks @jatin-bhateja for your help. I have confidence to integrate this now.

**Pull request closed without merging**
